# Codveda Technology-Level 1 Task 1
## Data Cleaning and preprocessing

### Project Objective

The objective of this project is to clean and preprocess the provided sentiment dataset by identifying and handling missing values, removing duplicates, standardizing inconsistent data formats, and ensuring that variables have appropriate data types for further analysis.

### Dataset Overview
The original dataset contains 732 rows and 15 columns. The dataset contains information about social media posts, including sentiment, timestamp, platform, hashtags, engagement metrics and country.

### Cleaning Process
The following preprocessing steps were performed:

1. Inspected the structure and data types of the dataset.
2. Checked for missing values.
3. Identified and removed unnecessary index columns.
4. Identified and removed duplicate records.
5. Removed leading and trailing whitespace from text-based columns.
6. Converted the `Timestamp` column to datetime format.
7. Validated the cleaned dataset.


### Final Result
After preprocessing, the dataset contained 710 rows and 13 columns, with no missing values or duplicate records.

## 1. Import Libraries

In [2]:
import pandas as pd

In [5]:
import os

os.listdir("Downloads")

['1769271568315.png',
 '1895 The Power Of Silence by Horatio W Dresser - Copy.pdf',
 '3) Sentiment dataset.csv',
 'Anaconda3-2025.12-2-Windows-x86_64.exe',
 'Build_the_Life_You_Want_The_Art_and_Science_of_Getting_Happier.pdf',
 'ChatGPT Installer.exe',
 'Code_Generated_Image.png',
 'desktop.ini',
 'DM Vasudevan - Textbook of Biochemistry For Medical Students, 6th Edition (2).pdf',
 'download (1).jfif',
 'download (2).jfif',
 'download (3).jfif',
 'download (4).jfif',
 'download (5).jfif',
 'download.jfif',
 'Find_Your_Why_A_Practical_Guide_for_Discovering_Purpose_for_You_.pdf',
 'FocusWriter_1.8.13.exe',
 'fortnite aura fortnitethumbnails image by @sssnipergamer.jfif',
 'GitHub - martinamx_Parks_and_Rec_Create_db_ This project is designed to create and manage a database for a Parks and Recreation department. It includes SQL scripts for database creation, table creation, data insertion, and analytica.html',
 'Glycemic_index_and_glycemic_load_Challenges_and_be.pdf',
 'LinkedIn_Banner (1)

## 2. Load The Dataset

The sentiment dataset provided for the internship was loaded into a pandas DataFrame for inspection and preprocessing.

In [6]:
df = pd.read_csv("Downloads/3) Sentiment dataset.csv")

## 3. Initial Data Inspection

The first five rows

In [7]:
df.head()

,Unnamed: 0.1,Unnamed: 0,Text,Sentiment,Timestamp,User,Platform,Hashtags,Retweets,Likes,Country,Year,Month,Day,Hour
0,0,0,Enjoying a beautiful day at the park! ...,Positive,2023-01-15 12:30:00,User123,Twitter,#Nature #Park,15.0,30.0,USA,2023,1,15,12
1,1,1,Traffic was terrible this morning. ...,Negative,2023-01-15 08:45:00,CommuterX,Twitter,#Traffic #Morning,5.0,10.0,Canada,2023,1,15,8
2,2,2,Just finished an amazing workout! 💪 ...,Positive,2023-01-15 15:45:00,FitnessFan,Instagram,#Fitness #Workout,20.0,40.0,USA,2023,1,15,15
3,3,3,Excited about the upcoming weekend getaway! ...,Positive,2023-01-15 18:20:00,AdventureX,Facebook,#Travel #Adventure,8.0,15.0,UK,2023,1,15,18
4,4,4,Trying out a new recipe for dinner tonight. ...,Neutral,2023-01-15 19:55:00,ChefCook,Instagram,#Cooking #Food,12.0,25.0,Australia,2023,1,15,19


In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 732 entries, 0 to 731
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0.1  732 non-null    int64  
 1   Unnamed: 0    732 non-null    int64  
 2   Text          732 non-null    object 
 3   Sentiment     732 non-null    object 
 4   Timestamp     732 non-null    object 
 5   User          732 non-null    object 
 6   Platform      732 non-null    object 
 7   Hashtags      732 non-null    object 
 8   Retweets      732 non-null    float64
 9   Likes         732 non-null    float64
 10  Country       732 non-null    object 
 11  Year          732 non-null    int64  
 12  Month         732 non-null    int64  
 13  Day           732 non-null    int64  
 14  Hour          732 non-null    int64  
dtypes: float64(2), int64(6), object(7)
memory usage: 85.9+ KB


## 4. Duplicate Records

Duplicate records were checked using pandas' `duplicated()` method.

The initial checked was performed before removing the unnecessary index columns. After those columns were removed, 22 genuine duplicate records were identified ad subsequently removed.

In [10]:
df.duplicated().sum()

np.int64(0)

In [11]:
df[["Unnamed: 0", "Unnamed: 0.1"]].head(10)

,Unnamed: 0,Unnamed: 0.1
0,0,0
1,1,1
2,2,2
3,3,3
4,4,4
5,5,5
6,6,6
7,7,7
8,8,8
9,9,9


In [12]:
df["Unnamed: 0"].equals(df["Unnamed: 0.1"])

False

In [13]:
df[["Unnamed: 0", "Unnamed: 0.1"]].tail(10)

,Unnamed: 0,Unnamed: 0.1
722,727,723
723,728,724
724,729,725
725,730,726
726,731,727
727,732,728
728,733,729
729,734,730
730,735,731
731,736,732


In [14]:
df["Unnamed: 0"].is_unique

True

In [15]:
df["Unnamed: 0.1"].is_unique

True

In [16]:
df[["Unnamed: 0", "Unnamed: 0.1"]].describe()

,Unnamed: 0,Unnamed: 0.1
count,732.000000,732.000000
mean,369.740437,366.464481
std,212.428936,211.513936
min,0.000000,0.000000
25%,185.750000,183.750000
50%,370.500000,366.500000
75%,553.250000,549.250000
max,736.000000,732.000000


In [17]:
df[["Unnamed: 0", "Unnamed: 0.1"]].tail(10)

,Unnamed: 0,Unnamed: 0.1
722,727,723
723,728,724
724,729,725
725,730,726
726,731,727
727,732,728
728,733,729
729,734,730
730,735,731
731,736,732


In [18]:
df[df["Unnamed: 0"] != df["Unnamed: 0.1"]][["Unnamed: 0", "Unnamed: 0.1"]].head(20)

,Unnamed: 0,Unnamed: 0.1
26,28,27
27,29,28
28,30,29
29,31,30
30,32,31
31,33,32
32,34,33
33,35,34
34,36,35
35,37,36


## 5. Removing unnecessary columns

The dataset contained two columns `Unnamed: 0` and `Unnamed: 0.1`, which were identified as redundant index-tracking columns rather than meaningful analytical variables.

Both columns were removed to retain only variables relevant to the analysis.

In [19]:
df = df.drop(columns=["Unnamed: 0", "Unnamed: 0.1"])

## Verification

The column names were checked after the removal to confirm that both unnecessary columns have been succesfully dropped.

In [20]:
df.columns

Index(['Text', 'Sentiment', 'Timestamp', 'User', 'Platform', 'Hashtags',
       'Retweets', 'Likes', 'Country', 'Year', 'Month', 'Day', 'Hour'],
      dtype='object')

In [21]:
df.dtypes

Text          object
Sentiment     object
Timestamp     object
User          object
Platform      object
Hashtags      object
Retweets     float64
Likes        float64
Country       object
Year           int64
Month          int64
Day            int64
Hour           int64
dtype: object

In [23]:
df["Timestamp"].head(10)

0    2023-01-15 12:30:00
1    2023-01-15 08:45:00
2    2023-01-15 15:45:00
3    2023-01-15 18:20:00
4    2023-01-15 19:55:00
5    2023-01-16 09:10:00
6    2023-01-16 14:45:00
7    2023-01-16 19:30:00
8    2023-01-17 08:00:00
9    2023-01-17 12:20:00
Name: Timestamp, dtype: object

In [25]:
pd.to_datetime(df["Timestamp"], errors="coerce").isna().sum()

np.int64(0)

## Data type standardization

The `Timestamp` column was initially stored as an object(text) data type. Since it represents date and time information, it was converted to pandas' datetime format.

The `Retweeta` and `Likes` columns were lso converted from floating-point values to integers because they contain whole number engagement counts.

In [26]:
df["Timestamp"] = pd.to_datetime(df["Timestamp"], errors="coerce")

In [27]:
df["Timestamp"].dtype

dtype('<M8[ns]')

In [28]:
df["Sentiment"].unique()

array([' Positive  ', ' Negative  ', ' Neutral   ', ' Anger        ',
       ' Fear         ', ' Sadness      ', ' Disgust      ',
       ' Happiness    ', ' Joy          ', ' Love         ',
       ' Amusement    ', ' Enjoyment    ', ' Admiration   ',
       ' Affection    ', ' Awe          ', ' Disappointed ',
       ' Surprise     ', ' Acceptance   ', ' Adoration    ',
       ' Anticipation ', ' Bitter       ', ' Calmness     ',
       ' Confusion    ', ' Excitement   ', ' Kind         ',
       ' Pride        ', ' Shame        ', ' Confusion ', ' Excitement ',
       ' Shame ', ' Elation       ', ' Euphoria      ', ' Contentment   ',
       ' Serenity      ', ' Gratitude     ', ' Hope          ',
       ' Empowerment   ', ' Compassion    ', ' Tenderness    ',
       ' Arousal       ', ' Enthusiasm    ', ' Fulfillment  ',
       ' Reverence     ', ' Compassion', ' Fulfillment   ', ' Reverence ',
       ' Elation   ', ' Despair         ', ' Grief           ',
       ' Loneliness     

In [29]:
df["Sentiment"].nunique()

279

In [30]:
df["Sentiment"].value_counts()

Sentiment
Positive           44
Joy                42
Excitement         32
Neutral            14
Contentment        14
                   ..
Adrenaline          1
Harmony             1
ArtisticBurst       1
Radiance            1
Elegance            1
Name: count, Length: 279, dtype: int64

In [32]:
df["Sentiment"].str.strip().nunique()

191

In [33]:
df["Sentiment"].str.strip().value_counts().head(20)

Sentiment
Positive       45
Joy            44
Excitement     37
Contentment    19
Neutral        18
Gratitude      18
Curiosity      16
Serenity       15
Happy          14
Nostalgia      11
Despair        11
Grief           9
Awe             9
Sad             9
Hopeful         9
Loneliness      9
Embarrassed     8
Acceptance      8
Confusion       8
Euphoria        7
Name: count, dtype: int64

## Text Standardization

Leading and trailing whitespace was removed from the relevant text-based columns to standardize categorical and textual values.

After cleaning, the number of unique values in the `Sentiment` column decreased from 279 to 191, indicating that whitespace inconsistencies had caused multple representations of the otherwise identical values. 

In [38]:
text_columns = ["Text", "Sentiment", "User", "Platform", "Hashtags", "Country"]

for column in text_columns:
 df[column] = df[column].str.strip()

In [39]:
df["Sentiment"].nunique()

191

In [40]:
df["Sentiment"].value_counts().head(20)

Sentiment
Positive       45
Joy            44
Excitement     37
Contentment    19
Neutral        18
Gratitude      18
Curiosity      16
Serenity       15
Happy          14
Nostalgia      11
Despair        11
Grief           9
Awe             9
Sad             9
Hopeful         9
Loneliness      9
Embarrassed     8
Acceptance      8
Confusion       8
Euphoria        7
Name: count, dtype: int64

In [41]:
df["Platform"].nunique()

3

In [42]:
df["Country"].nunique()

33

In [43]:
df["Platform"].value_counts()

Platform
Instagram    258
Twitter      243
Facebook     231
Name: count, dtype: int64

In [44]:
df["Country"].value_counts()

Country
USA               188
UK                143
Canada            135
Australia          75
India              70
Brazil             17
France             16
Japan              15
Germany            14
Italy              11
Spain               6
South Africa        6
Greece              5
Netherlands         4
Switzerland         3
Portugal            2
Austria             2
Belgium             2
Denmark             2
Czech Republic      2
Sweden              2
Colombia            1
Scotland            1
Kenya               1
Jamaica             1
Ireland             1
China               1
Norway              1
Cambodia            1
Maldives            1
Peru                1
Jordan              1
Thailand            1
Name: count, dtype: int64

In [45]:
df[["Retweets", "Likes"]].describe()

,Retweets,Likes
count,732.000000,732.000000
mean,21.508197,42.901639
std,7.061286,14.089848
min,5.000000,10.000000
25%,17.750000,34.750000
50%,22.000000,43.000000
75%,25.000000,50.000000
max,40.000000,80.000000


In [46]:
df[["Retweets", "Likes"]].apply(lambda x: (x % 1 != 0).sum())

Retweets    0
Likes       0
dtype: int64

In [48]:
df["Retweets"] = df["Retweets"].astype(int)
df["Likes"] = df["Likes"].astype(int)

In [49]:
df[["Retweets", "Likes"]].dtypes

Retweets    int64
Likes       int64
dtype: object

## 6. Missing Values Check

Missing values were checked across all columns using the `isnull().sum()` method.

the initial inspection showed that all columns contained 732 non-null values, indicating that there were no missing values in the original dataset.

In [50]:
df.isnull().sum()

Text         0
Sentiment    0
Timestamp    0
User         0
Platform     0
Hashtags     0
Retweets     0
Likes        0
Country      0
Year         0
Month        0
Day          0
Hour         0
dtype: int64

In [51]:
df.duplicated().sum()

np.int64(22)

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 732 entries, 0 to 731
Data columns (total 13 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Text       732 non-null    object        
 1   Sentiment  732 non-null    object        
 2   Timestamp  732 non-null    datetime64[ns]
 3   User       732 non-null    object        
 4   Platform   732 non-null    object        
 5   Hashtags   732 non-null    object        
 6   Retweets   732 non-null    int64         
 7   Likes      732 non-null    int64         
 8   Country    732 non-null    object        
 9   Year       732 non-null    int64         
 10  Month      732 non-null    int64         
 11  Day        732 non-null    int64         
 12  Hour       732 non-null    int64         
dtypes: datetime64[ns](1), int64(6), object(6)
memory usage: 74.5+ KB


In [53]:
df[df.duplicated(keep=False)].sort_values("Text").head(10)

,Text,Sentiment,Timestamp,User,Platform,Hashtags,Retweets,Likes,Country,Year,Month,Day,Hour
292,"A compassionate rain, tears of empathy falling...",Compassionate,2021-07-01 12:10:00,RainNurturer,Instagram,#Compassionate #TearsOfEmpathy,21,42,Canada,2021,7,1,12
278,"A compassionate rain, tears of empathy falling...",Compassionate,2021-07-01 12:10:00,RainNurturer,Instagram,#Compassionate #TearsOfEmpathy,21,42,Canada,2021,7,1,12
285,"A free spirit soaring on the wings of dreams, ...",Free-spirited,2020-06-10 10:05:00,DreamSoarer,Twitter,#FreeSpirit #WingsOfDreams,22,44,UK,2020,6,10,10
271,"A free spirit soaring on the wings of dreams, ...",Free-spirited,2020-06-10 10:05:00,DreamSoarer,Twitter,#FreeSpirit #WingsOfDreams,22,44,UK,2020,6,10,10
267,"A playful escapade in the carnival of life, ca...",Playful,2018-08-22 17:20:00,CarnivalDreamer,Facebook,#Playful #CarnivalEscapade,24,48,Australia,2018,8,22,17
281,"A playful escapade in the carnival of life, ca...",Playful,2018-08-22 17:20:00,CarnivalDreamer,Facebook,#Playful #CarnivalEscapade,24,48,Australia,2018,8,22,17
295,"A playful escapade in the carnival of life, ca...",Playful,2018-08-22 17:20:00,CarnivalDreamer,Facebook,#Playful #CarnivalEscapade,24,48,Australia,2018,8,22,17
272,"Bathed in the golden hues of gratefulness, a s...",Grateful,2022-04-01 18:30:00,SunsetAdmirer,Instagram,#Grateful #GoldenHues,19,38,Australia,2022,4,1,18
286,"Bathed in the golden hues of gratefulness, a s...",Grateful,2022-04-01 18:30:00,SunsetAdmirer,Instagram,#Grateful #GoldenHues,19,38,Australia,2022,4,1,18
291,Charting a course through the waves of hopeful...,Hopeful,2022-10-30 11:40:00,WaveNavigator,Twitter,#Hopeful #SailorOfDreams,17,34,Australia,2022,10,30,11


## Removing Duplicate Records

After the unnecessary index columns were removed,22 duplicate records were identified which were later confirmed and removed using `drop_duplicates()`.

In [54]:
df = df.drop_duplicates()

### Duplicate Validation

The dataset was checked again after removing the duplicates to confirm that no duplicate records remained.

In [55]:
df.duplicated().sum()

np.int64(0)

In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 710 entries, 0 to 731
Data columns (total 13 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Text       710 non-null    object        
 1   Sentiment  710 non-null    object        
 2   Timestamp  710 non-null    datetime64[ns]
 3   User       710 non-null    object        
 4   Platform   710 non-null    object        
 5   Hashtags   710 non-null    object        
 6   Retweets   710 non-null    int64         
 7   Likes      710 non-null    int64         
 8   Country    710 non-null    object        
 9   Year       710 non-null    int64         
 10  Month      710 non-null    int64         
 11  Day        710 non-null    int64         
 12  Hour       710 non-null    int64         
dtypes: datetime64[ns](1), int64(6), object(6)
memory usage: 77.7+ KB


## Export Cleaned Dataset

The cleaned dataset was exported as CSV  file for use in subsequent exploratory data analysis and other analytical tasks.

The original DataFrame index was excluded from the exported file to prevent the creation of an unnecessary index column.

In [58]:
df.to_csv("cleaned_sentiment_dataset.csv", index=False)

In [59]:
import os

os.getcwd()

'C:\\Users\\HP'

In [60]:
os.listdir()

['.anaconda',
 '.conda',
 '.continuum',
 '.idlerc',
 '.ipynb_checkpoints',
 '.ipython',
 '.jupyter',
 '.matplotlib',
 '.python_history',
 '.vscode',
 '3D Objects',
 'anaconda3',
 'AppData',
 'Application Data',
 'battery-report.html',
 'cleaned_sentiment_dataset.csv',
 'Codveda_Level_1_Task_1_Data_Cleaning.ipynb',
 'Codveda_Level_1_Task_2_EDA.ipynb',
 'Codveda_Level_1_Task_2_EDA.py',
 'Contacts',
 'Cookies',
 'Desktop',
 'DOC-20250519-WA0011..pptx',
 'Documents',
 'Downloads',
 'Favorites',
 'learning.py',
 'Links',
 'Local Settings',
 'MicrosoftEdgeBackups',
 'Music',
 'My Documents',
 'NetHood',
 'NTUSER.DAT',
 'ntuser.dat.LOG1',
 'ntuser.dat.LOG2',
 'NTUSER.DAT{b12ad8bb-dcde-11ef-b0d8-e86f3869a1f5}.TM.blf',
 'NTUSER.DAT{b12ad8bb-dcde-11ef-b0d8-e86f3869a1f5}.TMContainer00000000000000000001.regtrans-ms',
 'NTUSER.DAT{b12ad8bb-dcde-11ef-b0d8-e86f3869a1f5}.TMContainer00000000000000000002.regtrans-ms',
 'ntuser.ini',
 'OneDrive',
 'Pictures',
 'PrintHood',
 'Recent',
 'Saved Games',
 'Se